statsd/sender_resolving.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2012-2016 Eli Janssen
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file.
+
+package statsd
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+// ResolvingSimpleSender provides a socket send interface that re-resolves and
+// reconnects.
+type ResolvingSimpleSender struct {
+	// underlying connection
+	c net.PacketConn
+	// resolved udp address
+	ra *net.UDPAddr
+	// unresolved addr
+	ro string
+	// addr changed
+	ac bool
+	// ra mutex
+	ramx sync.RWMutex
+	// interval counter
+	ticker *time.Ticker
+	// done chan
+	doneChan chan bool
+	running  bool
+}
+
+// Send sends the data to the server endpoint.
+func (s *ResolvingSimpleSender) Send(data []byte) (int, error) {
+	// lock to guard against ra reconnection modification
+	s.ramx.RLock()
+	defer s.ramx.RUnlock()
+
+	if !s.running {
+		return 0, nil
+	}
+
+	// no need for locking here, as the underlying fdNet
+	// already serialized writes
+	n, err := s.c.(*net.UDPConn).WriteToUDP(data, s.ra)
+	if err != nil {
+		return 0, err
+	}
+	if n == 0 {
+		return n, errors.New("Wrote no bytes")
+	}
+	return n, nil
+}
+
+// Close closes the ResolvingSender
+func (s *ResolvingSimpleSender) Close() error {
+	// lock to guard against ra reconnection modification
+	s.ramx.Lock()
+	defer s.ramx.Unlock()
+	if !s.running {
+		return nil
+	}
+
+	s.running = false
+	s.doneChan <- true
+
+	err := s.c.Close()
+	return err
+}
+
+func (s *ResolvingSimpleSender) Reconnect() {
+	// lock to guard against s.running mutation
+	s.ramx.RLock()
+
+	if !s.running {
+		s.ramx.RUnlock()
+		return
+	}
+
+	// don't need to hold the lock while re-resolving addr
+	s.ramx.RUnlock()
+	fmt.Println("reconnecting!!")
+
+	// ro doesn't change, so no need to lock
+	ra, err := net.ResolveUDPAddr("udp", s.ro)
+	if err != nil {
+		// no good new address, so continue with old address?
+		return
+	}
+
+	// only need write lock when mutating ra
+	s.ramx.Lock()
+	defer s.ramx.Unlock()
+	// check running again, just to be sure nothing was terminated in the meantime...
+	if s.running {
+		s.ra = ra
```

Comment:
I don't think there would be much benefit.

1. the new ra is already resolved at this point.
2. If the new ra and the old ra resolve to the same IP, then we have to throw away one struct anyway.
3. doing the compare seems like it would take more cycles to dig into the UDPAddr structs, and then to do the byte comparison (UDPAdd.IP, which is a 4-byte (IPv4) or 16-byte (IPv6) slice), than just blindly doing the assignment.